Race condition in test_full_pipeline

While running the tests, I get an error sometimes that isn't always reproducible. It happened on my branch `schema-validation`, then as part of my testing also on `master`.

The assertion error is `assert 'aaa' != 'aaa'`, but can also be `assert 'bbb' != 'bbb'`.

    monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f18cc10a208>
    user_vk = 'JEAkEJqLbbgDRAtMm8YAjGp759Aq2qTn9eaEHUj2XePE'

    def test_full_pipeline(monkeypatch, user_vk):
        from bigchaindb.models import Transaction
        CONFIG = {
            'database': {
                'name': 'bigchain_test_{}'.format(os.getpid())
            },
            'keypair': {
                'private': '31Lb1ZGKTyHnmVK3LUMrAUrPNfd4sE2YyBt3UA4A25aA',
                'public': '4XYfCbabAWVUCbjTmRTFEu2sc3dFEdkse4r6X498B1s8'
            },
            'keyring': ['aaa', 'bbb'],
            'backlog_reassign_delay': 0.01
        }
        config_utils.set_config(CONFIG)
        b = Bigchain()
        outpipe = Pipe()

        original_txs = {}
        original_txc = []

        monkeypatch.setattr('time.time', lambda: 1)

        for i in range(100):
            tx = Transaction.create([b.me], [([user_vk], 1)])
            tx = tx.sign([b.me_private])
            original_txc.append(tx.to_dict())

            b.write_transaction(tx)
            original_txs[tx.id] = b.connection.run(r.table('backlog').get(tx.id))

        assert b.connection.run(r.table('backlog').count()) == 100

        monkeypatch.undo()

        pipeline = stale.create_pipeline(backlog_reassign_delay=1,
                                         timeout=1)
        pipeline.setup(outdata=outpipe)
        pipeline.start()

        # to terminate
        for _ in range(100):
            outpipe.get()

        pipeline.terminate()

        assert b.connection.run(r.table('backlog').count()) == 100
        reassigned_txs = list(b.connection.run(r.table('backlog')))

        # check that every assignment timestamp has increased, and every tx has a new assignee
        for reassigned_tx in reassigned_txs:
            assert reassigned_tx['assignment_timestamp'] > original_txs[reassigned_tx['id'] ]['assignment_timestamp']
    >       assert reassigned_tx['assignee'] != original_txs[reassigned_tx['id']]['assignee']
    E       assert 'aaa' != 'aaa'

    tests/pipelines/test_stale_monitor.py:116: AssertionError

--- a/tests/pipelines/test_stale_monitor.py
+++ b/tests/pipelines/test_stale_monitor.py
@@ -77,7 +77,6 @@ def test_full_pipeline(monkeypatch, user_pk):
     }
     config_utils.set_config(CONFIG)
     b = Bigchain()
-    outpipe = Pipe()
 
     original_txs = {}
     original_txc = []
@@ -96,9 +95,15 @@ def test_full_pipeline(monkeypatch, user_pk):
 
     monkeypatch.undo()
 
+    inpipe = Pipe()
+    # Each time the StaleTransactionMonitor pipeline runs, it reassigns
+    # all eligible transactions. Passing this inpipe prevents that from
+    # taking place more than once.
+    inpipe.put(())
+    outpipe = Pipe()
     pipeline = stale.create_pipeline(backlog_reassign_delay=1,
                                      timeout=1)
-    pipeline.setup(outdata=outpipe)
+    pipeline.setup(indata=inpipe, outdata=outpipe)
     pipeline.start()
 
     # to terminate

